### PR TITLE
[ON HOLD] When marking expression as size-like, substitute expr with fresh symint

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #117356
* __->__ #119043
* #118579

Partially addresses https://github.com/pytorch/pytorch/issues/93515

Removal of get_shape_groups because I don't think anyone was actually using it.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @aakhundov

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -516,13 +514,42 @@ def _constrain_range_for_size(a, min: Optional[int] = None, max: Optional[int] =
             "received min={min} and max={max}"
         )
 
+    shape_env = a.node.shape_env
+
+    vr = ValueRanges(min, max)
+    if free_unbacked_symbols(a.node.expr):
+        if (
+            not isinstance(a.node.expr, sympy.Symbol) and
+            not bound_sympy(a.node.expr, shape_env.var_to_range).issubset(vr)
```
Comment (lezcano) [inline torch/fx/experimental/symbolic_shapes.py:523]: I think we should take the intersection of both ranges

Comment (lezcano) [reply torch/fx/experimental/symbolic_shapes.py:523]: Actually that's the TODO you have a few lines down below.

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -516,13 +514,42 @@ def _constrain_range_for_size(a, min: Optional[int] = None, max: Optional[int] =
...
+            shape_env.var_to_range[u1] = vr
+            # Log this before the replacement occurs, so we can get the old
+            # a.node.expr
+            log.info("_constrain_range_for_size %s = %s [%s, %s]", a.node.expr, u1, min, max)
+            shape_env.complex_replacements[a.node.expr] = u1
+            shape_env._update_version_counter()
+            assert isinstance(a.node.expr, sympy.Symbol)
+    else:
+        assert isinstance(a.node.expr, sympy.Symbol), "constraining on non-symbol backed SymInt NYI"
+        # TODO: Could we do the unbacked replacement strategy here?  It's
```
Comment (lezcano) [inline torch/fx/experimental/symbolic_shapes.py:539]: Replace a backed symbol by an unbacked one? Why would we want that?

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -1796,11 +1823,18 @@ def _init(
         # Maps from sympy ints to expressions representing them
         # Populated from equality guards (i.e. a.shape[0] == b.shape[0])
         self.replacements: Dict[sympy.Symbol, sympy.Expr] = {}  #
+        # Complicated replacements from non-symbol expressions to other
+        # expressions.  These only ever involve unbacked SymInts on the LHS
+        self.complex_replacements: Dict[sympy.Expr, sympy.Expr] = {}
```
Comment (lezcano) [inline torch/fx/experimental/symbolic_shapes.py:1828]:
```suggestion
        self.complex_replacements: Dict[sympy.Expr, sympy.Symbol] = {}
```

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -3316,7 +3344,13 @@ def replace(expr, repl):
...
     @_lru_cache
     def replace(self, expr: "sympy.Expr") -> "sympy.Expr":
-        replacements = {s: self._find(cast(sympy.Symbol, s)) for s in expr.free_symbols}
+        replacements = {}
+        for s in expr.free_symbols:
+            new_s = self._find(cast(sympy.Symbol, s))
+            if new_s != s:
+                replacements[s] = new_s
+        if free_unbacked_symbols(expr):
+            replacements.update(self.complex_replacements)
```
Comment (lezcano) [inline torch/fx/experimental/symbolic_shapes.py:3353]: We should also do replacements of `self.replacements` in `self.complex_replacements`. We can do that while adding a new `complex_replacement`. Or is it an invariant that these replacements have already been applied? If so, can you assert that here?

**Comment by lezcano** [comment]
>We can in fact answer this question by performing a substitution of u1 back to 3 * u0. But it's not obvious to me how exactly we can do this. 

The current way is a bit too hacky. What about: We know the order in which the size-like uints were created. We look at the earliest uint used in the whole expression and we perform the substitutions up to that point. Note that we would need to apply them one by one starting at the last one (sort of like peeling out the expressions). We discussed some sort of algorithm like this in our call at some point.

File: torch/fx/experimental/symbolic_shapes.py (hunk 2 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:539]: Replace backed with backed. And for the same reason: you can reliably refine the range afterwards.

**Comment by ezyang** [comment]
We're putting this on ice for now

**Comment by ezyang** [comment]
https://fb.workplace.com/groups/6829516587176185/posts/6853529854774858/ may be a concrete case of this mattering

Blurb

```
	
                sz = end_offset - start_offset

                torch._check(start_offset <= self._values.size(0))
                torch._check(sz <= self._values.size(0))
                torch._check(sz != 0)
                torch._check(sz != -1)
                torch._check(sz != 1)
                torch._check_is_size(start_offset)
                torch._check_is_size(sz)

                lengths_start = self.lengths_offset_per_key()[start]
                lengths_sz = self.lengths_offset_per_key()[end] - lengths_start

                _lengths = torch.narrow(self.lengths(), 0, lengths_start, lengths_sz)
```

**Pull request closed without merging**